File: app/javascript/channels/room_channel.js
import consumer from "./consumer"

$(document).on('turbolinks:load', function () {

  const messages = $('#messages')

  const room = consumer.subscriptions.create( {
                  channel: 'RoomChannel',
                  room: $('#messages').data('room_id'),
              }, {
    connected() {
      // un used
    },

    disconnected() {
      // un used
    },

    received(data) {
      messages.append(data['message'])
    },

    speak: function (message) {
      return this.perform('speak', {message: message})
    },

  })

  $(document).on('keypress', '[data-behavior~=room_speaker]', function(event) {
    if (event.keyCode === 13) {
      room.speak(event.target.value);
      event.target.value = '';
      return event.preventDefault();
    }
  });

})

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  def subscribed
    stream_from "room_channel_#{params['room']}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    record = Message.create! chat: data['message'], user_id: current_user.id, room_id: params['room']
    ActionCable.server.broadcast("room_channel_#{params['room']}", {message: render_message(record)})
  end

  private

  def render_message(record)
    ApplicationController.renderer.render(partial: 'messages/message', locals: {message: record})
  end

end
